src/Glide64/3dmath.cpp

```diff
@@ -190,16 +190,37 @@ void InverseTransformVectorC (float *src, float *dst, float mat[4][4])
 
 void MulMatricesC(float m1[4][4],float m2[4][4],float r[4][4])
 {
-  for (int i=0; i<4; i++)
-  {
-    for (int j=0; j<4; j++)
+    float row[4][4];
+    register unsigned int i, j;
+
+    for (i = 0; i < 4; i++)
+        for (j = 0; j < 4; j++)
+            row[i][j] = m2[i][j];
+    for (i = 0; i < 4; i++)
     {
-      r[i][j] = m1[i][0] * m2[0][j] +
-                m1[i][1] * m2[1][j] +
-                m1[i][2] * m2[2][j] +
-                m1[i][3] * m2[3][j];
+        float leftrow[4], destrow[4];
```

Comment:
I like it, too.  Unfortunately, I just checked, and it did not provide the same assembly language results.

It added dynamic shuffle instruction code due to the new `k` variable, where shuffling one element across all four single-precision floating-point elements of the 128-bit xmm was no longer as fixated when using a loop variable instead of a literal 0, 1, 2, 3 for array index.

This is unfortunate I think, since we seem to agree that this option C) at the bottom of my reply is more readable.  Maybe it will go away in a future version of the GCC compiler, and you can use it without any performance loss whatsoever.  I sometimes observe that the current version of GCC is not good at auto-vectorizing but that the upcoming release of GCC will have such volatility fixed; you can try that if you want.  I am currently compiling on 32-bit Linux with GCC 4.8.2; you might have a better version for this particular limit.

Here is the exact,  full C code I am using to debug this btw:
```float summand[4][4], leftrow[4], row[4][4];

int main(void)
{
    register unsigned int j, k;

    for (k = 0; k < 4; k++)
        for (j = 0; j < 4; j++)
            summand[k][j] = leftrow[k] * row[k][j];
    return 0;
}```

Compiled with:
`gcc -S -O3 -msse2 -o /home/no/m64/main.s /home/no/m64/main.c`
Omit the `-msse2` flag if compiling in GCC for 64-bit x86 code.
Check `main.s` for the assembly language results of the code.

It was the same with the first two methods I gave, but the third method for all its readability and flexibility seems less rigid to the compiler's ability to guarantee single-move shuffling in the SHUFPS op-code.